woocommerce/woocommerce
# Description Generation - Security + Sanitization

### Description
Add some basic security to the Woo AI Product Description endpoint + client code.
Some questions I was attempting to answer (some already solved previously):

- [x] Can a random agent use the product description REST API endpoint? NO - we have a permissions check to make sure they're logged in and are able to edit products.
- [x] Do we need a nonce to increase the security of the product description generator? NO - since this is done within the WP Admin area, a nonce is already generated and sent in the header.
- [x] If the text completions service that we integrate with returns malicious code, do we dump that into the product editor? NO - we strip out unwanted tags, leaving some HTML tags used for WP posts.
- [x] Can someone enter a gigantic product title and generate a description, massively bloating the cost of text completions due to the number of tokens in the prompt? NO - We now truncate the product title string used for the prompt.

### Changes proposed in this Pull Request:

- [x] ~Generate and verify a nonce on the product editor page for generating a product description using apiFetch middleware.~ Not necessary since this implementation is all done within wp admin; a nonce will automatically be sent using the X-WP-Nonce header and validated automatically by our REST API endpoint.
- [x] Use wp_kses_post() on the messages returned from the OpenAI endpoint to ensure we are not returning tags that can execute arbitrary code.
- [x] Strip all tags from the arguments passed into the REST API endpoint _before_ they're sent to the OpenAI endpoint.
- [x] Truncate the Product Title string injected into the prompt to prevent bloating the prompt since the title input doesn't have a limit set. I chose a somewhat opinionated 150 character count limit as this appears to be the limit for Google Shopping product titles.

This is by no means comprehensive as I'm sure things will change as we move the endpoints off of the site and to WPCOM/Jetpack but it's surely better to have this in place than nothing. 
Feel free to point out any vectors I'm missing here 👍 

Closes #38259 



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to the product editor and ensure that the "Happy Path" works by generating a product description.
2. Add a super long title (say, 200+ characters), generate a product description and inspect the prompt in the request to ensure that it truncated the title within the prompt but left the rest of the prompt intact.
3. Mock the return of the OpenAI text completions API request and set the content of the generated response to have <script> tags, observe that they're removed when returned back from our own endpoint.
